### ISSUE TITLE
Exported GPX distance, average speed and elevation gain are inaccurate

### ISSUE THREAD

**Issue by robsonj** [nachmore/unBand#65]
I exported a bike activity in GPX format and then imported into Garmin Connect. The stated average speed is 35,207.2mph (I am superman!), the distance was -2.424.05miles. Elevation gain was also 1,388ft when in reality it was probably about 100-200ft.

Heartrate and gps all look good.

Link to the uploaded activity from the gpx: https://connect.garmin.com/activity/730962487

Link to the gpx file: https://www.dropbox.com/s/53zl2h5tnjdhszy/Biking_2519753374335512497.gpx?dl=0

Summary
Distance:	-2,424.05 mi
Time:	41:21
Avg Speed:	35,207.2 mph
Elevation Gain:	1,388 ft

Timing
Time:	41:21
Moving Time:	36:23
Elapsed Time:	41:21
Avg Speed:	35,207.2 mph
Avg Moving Speed:	40,013.3 mph
Max Speed:	2,156,628.2 mph

![image](https://cloud.githubusercontent.com/assets/1464175/6881349/f493d36e-d529-11e4-9609-6fdf111b5a58.png)

Cheers
Jonathan

**Comment by nachmore** [comment]
Seriously, you should change your name to Clark Kent!

Thanks for the bug report - I'll take a look...

**Comment by manzanotti** [comment]
You'll probably find that the first and last entries in the GPX file have a lat/long of 0/0. There will also be 0/0 entries any time you pause the Band during a ride.

I've downloaded the project to see if I could fix this, but I can't get it to build, so that's not going to happen!

**Comment by robsonj** [comment]
Yep you are correct manzanotti, that's the issue. Same here though. I cant get the solution to build

**Comment by danai84** [comment]
I have same issue, when I export runs or bike rides as a GPX file the distances are in their thousands of miles. This is not consistent though, as some of my exports work.

I have asked app to show list of activities, selected a file with a gps record, then clicked export GPX file. This one worked  Run_20150417_0647_map. This one didn't, showing a crazy distance covered Biking_20150420_0907_map.

here's the file source: https://onedrive.live.com/redir?resid=6B9C24DD213F835A!50743&authkey=!AMg1lbDzEdTsYuA&ithint=folder%2cgpx

**Comment by robsonj** [comment]
After getting unband to compile I was able to capture the json response data for the bike event I detailed above...

https://www.dropbox.com/s/pufwxqg5mcx6lfv/bikedata.json?dl=0

**Comment by robsonj** [comment]
With a quick fix to just filter the lat longs of 0 & 0 - which for mine was just the first one, everything looks good...

https://www.dropbox.com/s/cjr45i8wwuilycd/gpx%20fix.jpg?dl=0

Need to write unit tests for the fix and see if nachmore is happy for me to contribute a fix...

**Comment by manzanotti** [comment]
How did you get it to build in the end? Feels like there should be a better fix than just ignoring 0 lat/longs, especially with regard to pausing during a ride, would like to have a look at seeing if I can find one!

**Comment by robsonj** [comment]
For me there wher esome invalid project references that neded cleaning up...

https://pbs.twimg.com/media/CDpKmIrWYAIq9GT.jpg:large

in some places the band dll's had references to renamed versions (Microsoft.Band.unband.dll etc) other places they didn't. The bootstrap project does not or no longer performs a rename, so I fixed those. There was also a minor code change due to a ToString() on a missing property - I assume the property got removed from the public API in the band dll's.

The most annoying one was a missing reference to System.IO.dll which threw me as I thought that namespace was in the mscorlib.dll, but it turns out that unBand.CargoClientExtender.BluetoothDeviceTransport implements Microsoft.Band.IDeviceTransport from the Microsoft.Band.dll which has a reference to a portable dll as seen here...

https://pbs.twimg.com/media/CDpLGWSWAAIwr7v.jpg:large

Once I added in that reference I was good.

Only other thing I did was disable code signing - purely just for compiling purposes, something I wouldn't commit that change.

**Comment by robsonj** [comment]
Regarding the 0 lat/long. My fix was just a quick 5 minute see if it works before I called it a night. But in my case, filtering them out is probably the only valid solution as it was the very first map point that had 0 lat/long. I suspect the GPS hadn't locked yet. For points in the middle, yea maybe trying to approximate would be better, but I haven't seen that myself yet.

**Comment by robsonj** [comment]
I sumbitted a pull request to merge in the build fixes.

**Comment by robsonj** [comment]
Submitted Pull req with basic filtering functionality...

https://github.com/nachmore/unBand/pull/74
